ubcpi/answer_pool.py

```diff
@@ -0,0 +1,315 @@
+    num_responses = len(options) \
+        if 'num_responses' not in algo or algo['num_responses'] == "#" \
+        else int(algo['num_responses'])
+
+    if algo['name'] == 'simple':
+        return get_other_answers_simple(answers, seeded_answers, get_student_item_dict, num_responses)
+    elif algo['name'] == 'random':
+        return get_other_answers_random(answers, seeded_answers, get_student_item_dict, num_responses)
+    else:
+        raise UnknownChooseAnswerAlgorithm()
+
+
+def get_other_answers_simple(answers, seeded_answers, get_student_item_dict, num_responses):
+    """
+    Get answers from others with simple algorithm, which picks one answer for each option.
+
+    Args:
+        see `get_other_answers`
+        num_responses (int): the number of responses to be returned. This value may not be
+            respected if there is not enough answers to return
+
+    Returns:
+        dict: answers based on the selection algorithm
+    """
+    ret = []
+    # clean up answers so that all keys are int
+    answers = {int(k): v for k, v in answers.items()}
+    total_in_pool = len(seeded_answers)
+    pool = convert_seeded_answers(seeded_answers)
+    student_id = get_student_item_dict()['student_id']
+    # merge the dictionaries in the answer dictionary
+    for key in answers:
+        total_in_pool += len(answers[key])
+        # if student_id has value, we assume the student just submitted an answer. So removing it
+        # from total number in the pool
+        if student_id in answers[key].keys():
+            total_in_pool -= 1
+        if key in pool:
+            pool[key].update(answers[key].items())
+        else:
+            pool[key] = answers[key]
+
+    # remember which option+student_id is selected, so that we don't have duplicates in the result
+    selected = []
+
+    # loop until we have enough answers to return
+    while len(ret) < min(num_responses, total_in_pool):
+        for option, students in pool.items():
+            student = student_id
+            i = 0
+            while (student == student_id or i > 100) and (str(option) + student) not in selected:
+                # retry until we got a different one or after 100 retries
+                # we are suppose to get a different student answer or a seeded one in a few tries
+                # as we have at least one seeded answer for each option in the algo. And it is not
+                # suppose to overflow i order to break the loop
+                student = random.choice(students.keys())
+                i += 1
+            selected.append(str(option)+student)
+            if student.startswith('seeded'):
+                # seeded answer, get the rationale from local
+                rationale = students[student]
+            else:
+                student_item = get_student_item_dict(student)
+                submission = sas_api.get_answers_for_student(student_item)
```

Comment:
> maybe I can propose a bulk query API

Yes - that's a great idea, and one of the ideas Dave and I discussed
yesterday. Would you be interested in implementing that? If so I'd like to
connect you up with Dave in advance of project execution, to make sure that
you don't have to do rounds of code refactoring.

On Fri, Sep 11, 2015 at 4:03 AM, Pan Luo <notifications@github.com> wrote:

> In ubcpi/answer_pool.py
> <https://github.com/ubc/ubcpi/pull/27#discussion_r39249259>:
>
> > +            student = student_id
> > +            i = 0
> > +            while (student == student_id or i > 100) and (str(option) + student) not in selected:
> > +                # retry until we got a different one or after 100 retries
> > +                # we are suppose to get a different student answer or a seeded one in a few tries
> > +                # as we have at least one seeded answer for each option in the algo. And it is not
> > +                # suppose to overflow i order to break the loop
> > +                student = random.choice(students.keys())
> > +                i += 1
> > +            selected.append(str(option)+student)
> > +            if student.startswith('seeded'):
> > +                # seeded answer, get the rationale from local
> > +                rationale = students[student]
> > +            else:
> > +                student_item = get_student_item_dict(student)
> > +                submission = sas_api.get_answers_for_student(student_item)
>
> @sarina <https://github.com/sarina> actually those API calls will not
> happen on the block render. It only happens when the student submit his/her
> answer first time. Xblock will then pick the answers from the pool and
> retrieve the rationale through submission API.
>
> As for querying the submission API, maybe I can propose a bulk query API
> to retrieve a list of submissions so that we can reduce the number of the
> queries.
>
> For now, I think we can limit to 8 or 10. I'll update the code tomorrow.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ubc/ubcpi/pull/27/files#r39249259>.
>